feature/experimental/p4rt/ate_tests/performance_test/performance_test.go

```diff
@@ -0,0 +1,950 @@
+	}
+
+	atePort2 = attrs.Attributes{
+		Name:    "atePort2",
+		IPv4:    "192.0.2.6",
+		IPv6:    "2001:db8::6",
+		IPv4Len: ipv4PrefixLen,
+		IPv6Len: ipv6PrefixLen,
+	}
+)
+
+type testArgs struct {
+	ctx          context.Context
+	leader       *p4rt_client.P4RTClient
+	dut          *ondatra.DUTDevice
+	ate          *ondatra.ATEDevice
+	top          *ondatra.ATETopology
+	gdppacketIO  PacketIO
+	lldppacketIO PacketIO
+	trpacketIO   PacketIO
+}
+
+type PacketIOPacket struct {
+	TTL            *uint8
+	SrcMAC, DstMAC *string
+	EthernetType   *uint32
+	HopLimit       *uint8
+}
+
+type PacketIO interface {
+	GetPacketTemplate() *PacketIOPacket
+	GetEgressPort() string
+	GetIngressPort() string
+}
+
+type LLDPPacketIO struct {
+	PacketIOPacket
+	IngressPort string
+}
+
+type GDPPacketIO struct {
+	PacketIOPacket
+	IngressPort string
+}
+
+type TraceroutePacketIO struct {
+	PacketIOPacket
+	IngressPort string
+	EgressPort  string
+}
+
+// programTableEntry programs or deletes p4rt table entry based on delete flag.
+func programTableEntry(client *p4rt_client.P4RTClient, delete bool) error {
+	return client.Write(&p4_v1.WriteRequest{
+		DeviceId:   deviceID,
+		ElectionId: &p4_v1.Uint128{High: uint64(0), Low: electionID},
+		Updates: p4rtutils.ACLWbbIngressTableEntryGet(
+			getTableEntry(delete),
+		),
+		Atomicity: p4_v1.WriteRequest_CONTINUE_ON_ERROR,
+	})
+}
+
+// Send ATE traffic, stop and collect total packet tx from ATE source port.
+func testTraffic(t *testing.T, ate *ondatra.ATEDevice, flows []*ondatra.Flow, srcEndPoint *ondatra.Interface, duration int) int {
+	t.Helper()
+	for _, flow := range flows {
+		flow.WithSrcEndpoints(srcEndPoint).WithDstEndpoints(srcEndPoint)
+	}
+	ate.Traffic().Start(t, flows...)
+	time.Sleep(time.Duration(duration) * time.Second)
+	ate.Traffic().Stop(t)
+
+	outPkts := gnmi.GetAll(t, ate, gnmi.OC().FlowAny().Counters().OutPkts().State())
+	total := 0
+	for _, count := range outPkts {
+		total += int(count)
+	}
+	return total
+}
+
+// sendPackets GDP/LLDP/Traceroute sends out packets via PacketOut message in StreamChannel.
+func sendPackets(t *testing.T, client *p4rt_client.P4RTClient, packet *p4_v1.PacketOut, packetCount int, timer float64) {
+
+	for i := 0; i < packetCount; i++ {
+		if err := client.StreamChannelSendMsg(
+			&streamName, &p4_v1.StreamMessageRequest{
+				Update: &p4_v1.StreamMessageRequest_Packet{
+					Packet: packet,
+				},
+			}); err != nil {
+			t.Errorf("There is error seen in Packet Out. %v, %s", err, err)
+		}
+		time.Sleep(time.Duration(timer) * time.Millisecond)
+	}
+}
+
+// decodePacket decodes L2 header in the packet and returns TTL, packetData[14:0] to remove first 14 bytes of Ethernet header.
+func decodeIPpacket(t *testing.T, packetData []byte) uint8 {
+	t.Helper()
+	packet := gopacket.NewPacket(packetData, layers.LayerTypeEthernet, gopacket.Default)
+	etherHeader := packet.Layer(layers.LayerTypeEthernet)
+	if etherHeader != nil {
+		header, _ := etherHeader.(*layers.Ethernet)
```

Comment:
should shouldn't ignore the ok return value from the type assertion. Either(1) leave out the second return value, so the code panics if the header is not the right type:

`header := etherHeader.(*layers.Ethernet)`

or (2) capture the ok value and fail the test:

```
header, ok := etherHeader.(*layers.Ethernet)
if !ok {
  t.Fatalf(...)
}
```